src/uct/ugni/smsg/ugni_smsg_iface.c

```diff
@@ -0,0 +1,432 @@
+  gni_return_t ugni_rc;
+  gni_cq_entry_t event_data;
+  uct_ugni_smsg_desc_t message_data;
+  uct_ugni_smsg_desc_t *message_pointer;
+
+  if(iface->super.outstanding == 0){
+    return;
+  }
+
+  ugni_rc = GNI_CqGetEvent(iface->super.local_cq, &event_data);
+  if(ugni_rc == GNI_RC_NOT_DONE){
+    return;
+  }
+
+  if((GNI_RC_SUCCESS == ugni_rc && !event_data) || GNI_CQ_OVERRUN(event_data)){
+    /* TODO: handle overruns */
+    ucs_error("Error posting data. CQ overrun = %d", (int)GNI_CQ_OVERRUN(event_data));
+    return;
+  }
+
+  message_data.msg_id = GNI_CQ_GET_MSG_ID(event_data);
+  message_pointer = sglib_hashed_uct_ugni_smsg_desc_t_find_member(iface->smsg_list,&message_data);
+  message_pointer->ep->outstanding--;
+  iface->super.outstanding--;
+  sglib_hashed_uct_ugni_smsg_desc_t_delete(iface->smsg_list,message_pointer);
+  ucs_mpool_put(message_pointer);
+}
+
+static void process_mbox(uct_ugni_smsg_iface_t *iface, uct_ugni_smsg_ep_t *ep){
+  ucs_status_t status;
+  uint8_t tag;
+  void *data_ptr;
+  gni_return_t ugni_rc;
+  uct_ugni_smsg_header_t *header;
+  void *user_data;
+
+  pthread_mutex_lock(&uct_ugni_global_lock);
+
+  while(1){
+    tag = GNI_SMSG_ANY_TAG;
+    ugni_rc = GNI_SmsgGetNextWTag(ep->super.ep, (void **)&data_ptr, &tag);
+
+    /* Yes, GNI_RC_NOT_DONE means that you're done with the smsg mailbox */
+    if(GNI_RC_NOT_DONE == ugni_rc){
+      pthread_mutex_unlock(&uct_ugni_global_lock);
+      return;
+    }
+
+    if(GNI_RC_SUCCESS != ugni_rc){
+      ucs_error("Unhandled smsg error: %s %d", gni_err_str[ugni_rc], ugni_rc);
+      pthread_mutex_unlock(&uct_ugni_global_lock);
+      return;
+    }
+
+    if(NULL == data_ptr){
+      ucs_error("Empty data pointer in smsg.");
+      pthread_mutex_unlock(&uct_ugni_global_lock);
+      return;
+    }
+
+    header = (uct_ugni_smsg_header_t *)data_ptr;
+    user_data = (void *)(header + 1);
+    void *user_desc = iface->user_desc+1;
+
+    uct_iface_trace_am(&iface->super.super, UCT_AM_TRACE_TYPE_RECV,
+                       tag, user_data, header->length, "RX: AM");
+
+    status = uct_iface_invoke_am(&iface->super.super, tag, user_data,
+                                 header->length, user_desc);
+
+    if(status != UCS_OK){
+        uct_recv_desc_iface(user_desc) = &iface->super.super.super;
+        UCT_TL_IFACE_GET_TX_DESC(&iface->super.super, &iface->free_desc,
+                                 iface->user_desc, iface->user_desc = NULL);
+    }
+
+    GNI_SmsgRelease(ep->super.ep);
+  }
+}
+
+static void progress_remote_cq(uct_ugni_smsg_iface_t *iface){
+  gni_return_t ugni_rc;
+  gni_cq_entry_t event_data;
+  uct_ugni_ep_t tl_ep;
+  uct_ugni_ep_t *ugni_ep;
+  uct_ugni_smsg_ep_t *ep;
+
+  ugni_rc = GNI_CqGetEvent(iface->remote_cq, &event_data);
+
+  if(GNI_RC_NOT_DONE == ugni_rc){
+    return;
+  }
+
+  if (GNI_RC_SUCCESS != ugni_rc || GNI_CQ_OVERRUN(event_data)) {
+    ucs_error("GNI_CqGetEvent falied. Error status %s %d ",
```

Comment:
what's the strategy for recovering from CQ overrun status?  Production code will need to have a plan B to handle this.  Maybe its being handled at a higher level?